models/migrations/v1_22/v283.go

```diff
@@ -0,0 +1,350 @@
+}
+
+// detectLicenseByEntry returns the licenses detected by the given tree entry
+func detectLicenseByEntry(file *git.TreeEntry) ([]string, error) {
+	if file == nil {
+		return nil, nil
+	}
+
+	blob := file.Blob()
+	content, err := blob.GetBlobContent(setting.UI.MaxDisplayFileSize)
+	if err != nil {
+		return nil, fmt.Errorf("GetBlobAll: %w", err)
+	}
+	return detectLicense(content), nil
+}
+
+// detectLicense returns the licenses detected by the given content buff
+func detectLicense(content string) []string {
+	if len(content) == 0 {
+		return nil
+	}
+	if classifier == nil {
+		log.Error("detectLicense: license classifier is null.")
+		return nil
+	}
+
+	matches, err := classifier.MatchFrom(strings.NewReader(content))
+	if err != nil {
+		log.Error("licenseclassifier.MatchFrom: %v", err)
+		return nil
+	}
+	if len(matches.Matches) > 0 {
+		results := make(container.Set[string], len(matches.Matches))
+		for _, r := range matches.Matches {
+			if r.MatchType == "License" && !results.Contains(r.Variant) {
+				results.Add(r.Variant)
+			}
+		}
+		return results.Values()
+	}
+	return nil
+}
+
+func ConvertLicenseName(name string) string {
+	if sameLicenses == nil {
+		return name
+	}
+
+	v, ok := sameLicenses[name]
+	if ok {
+		return v
+	}
+	return name
+}
+
+func initClassifier() error {
+	data, err := options.AssetFS().ReadFile("", "sameLicenses")
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(data, &sameLicenses)
+	if err != nil {
+		return err
+	}
+
+	// threshold should be 0.84~0.86 or the test will be failed
+	classifier = licenseclassifier.NewClassifier(.85)
+	licenseFiles, err := options.AssetFS().ListFiles("license", true)
+	if err != nil {
+		return err
+	}
+
+	licenseNameCount := make(map[string]int)
+	if len(licenseFiles) > 0 {
+		for _, licenseFile := range licenseFiles {
+			data, err := options.License(licenseFile)
+			if err != nil {
+				return err
+			}
+			licenseName := ConvertLicenseName(licenseFile)
+			licenseNameCount[licenseName]++
+			if licenseNameCount[licenseName] > 1 {
+				continue
+			}
+			classifier.AddContent("License", licenseFile, licenseName, data)
+		}
+	}
+	return nil
+}
+
+func AddRepositoryLicenses(x *xorm.Engine) error {
+	type Repository struct {
+		ID            int64 `xorm:"pk autoincr"`
+		OwnerName     string
+		Name          string `xorm:"INDEX NOT NULL"`
+		DefaultBranch string
+	}
+
+	type RepoLicense struct {
+		ID          int64 `xorm:"pk autoincr"`
+		RepoID      int64 `xorm:"UNIQUE(s) INDEX NOT NULL"`
+		CommitID    string
+		License     string             `xorm:"VARCHAR(50) UNIQUE(s) INDEX NOT NULL"`
+		CreatedUnix timeutil.TimeStamp `xorm:"INDEX CREATED"`
+		UpdatedUnix timeutil.TimeStamp `xorm:"INDEX UPDATED"`
+	}
+
+	if err := x.Sync(new(RepoLicense)); err != nil {
+		return err
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	repos := make([]*Repository, 0)
+	if err := sess.Find(&repos); err != nil {
```

Comment:
There will be a lot of repos here for bigger instances. We never merged #18871 because the migration could take some time. This migration will take even more time.

Reply:
I don't think we should really stop merging big features because of these few big instances. Yes, it will take time, but compared to the number of Codeberg-like instances there are numerous smaller ones that would really benefit from it.

An alternative would be to hide it behind a doctor command, but many users don't even know how to execute them.

Reply:
Don't get me wrong. I want the features too and don't like #18871 did not get merged.

Reply:
🤔 Can we only add table in migrations, and add a `Sync Licences` task or something else in Admin page?
Then user can do this later by themselves.